pkg/eth/transformer.go

```diff
@@ -0,0 +1,355 @@
+// VulcanizeDB
+// Copyright © 2019 Vulcanize
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package eth
+
+import (
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/ethereum/go-ethereum/statediff"
+	node "github.com/ipfs/go-ipld-format"
+	"github.com/jmoiron/sqlx"
+	"github.com/multiformats/go-multihash"
+	"github.com/sirupsen/logrus"
+
+	"github.com/vulcanize/ipld-eth-indexer/pkg/ipfs/ipld"
+	"github.com/vulcanize/ipld-eth-indexer/pkg/postgres"
+	"github.com/vulcanize/ipld-eth-indexer/pkg/shared"
+)
+
+// Transformer interface to allow substitution of mocks for testing
+type Transformer interface {
+	Transform(workerID int, payload statediff.Payload) (int64, error)
+}
+
+// StateDiffTransformer satisfies the Transformer interface for ethereum statediff objects
+type StateDiffTransformer struct {
+	chainConfig *params.ChainConfig
+	indexer     *CIDIndexer
+}
+
+// NewStateDiffTransformer creates a pointer to a new PayloadConverter which satisfies the PayloadConverter interface
+func NewStateDiffTransformer(chainConfig *params.ChainConfig, db *postgres.DB) *StateDiffTransformer {
+	return &StateDiffTransformer{
+		chainConfig: chainConfig,
+		indexer:     NewCIDIndexer(db),
+	}
+}
+
+// Transform method is used to process statediff.Payload objects
+// It performs the necessary data conversions and database persistence
+func (sdt *StateDiffTransformer) Transform(workerID int, payload statediff.Payload) (int64, error) {
+	// Unpack block rlp to access fields
+	block := new(types.Block)
+	if err := rlp.DecodeBytes(payload.BlockRlp, block); err != nil {
+		return 0, err
```

Comment:
don't we need to add some log warning message to the cases when we couldn't decode RLP?

Reply:
The service calling the Transform method will provide some additional context to the error and log it. Since there are two rlp decodes in the method I think it would be a good idea to add some additional context at this level to differentiate between them- will do that.